### ISSUE TITLE
[REQ]Is that possible to support chinese character password input ?

### ISSUE THREAD

**Issue by liushulun** [aonez/Keka#656]
Hi all, I'm so glad to have this utility on mac. But I'm wondering that, although compress / decompress with chinese character password is supported, I still need to copy the character from other place and paste into keka, it's somehow not convenient enough. And sometimes if I forget what I've copied, I'll have to check the content, that makes it low efficiency.

Looking forward to your reply~

**Comment by aonez** [comment]
Hi @liushulun! I'm sorry I'm not familiar with the way Chinese characters are typed. What you need that Keka is lacking to write those characters? Maybe you can share a screencast showing what's needed?

**Comment by Gimmy1206** [comment]
![截圖 2020-10-28 11 32 29](https://user-images.githubusercontent.com/3322530/97387719-86b3c980-1911-11eb-9449-c7c67a18805a.png)

Keka always get error message when the password with Chinese character.
but use BetterZip is fine.

**Comment by aonez** [comment]
Please @Gimmy1206 open a separate issue if you have trouble extracting a file.
Anyway it should work, just check those RAR a 7Z test files with password _汉字_: [Test.zip](https://github.com/aonez/Keka/files/5450415/Test.zip)

**Comment by Gimmy1206** [comment]
test.rar and test.7z all work.
I don't understand what happened. maybe is the password too long?
you can test this RAR with password【x-o-x.org】【44FB985C08_by_MEGAFUNPRO.COM】【17C8472BFA_by_OKFUN.ORG】【FCBZONEBDD4F0A112_by_FCBZONE.ORG】【HG4S83816NBVD_by_peru-boundary.com】【FASTJHM63C6BFA69NPG_by_FastZone.ORG】【倉庫 @http://txbpyc.com】: 
[test.zip](https://github.com/aonez/Keka/files/5451989/test.zip)

**Comment by gingerbeardman** [comment]
the password is this whole long line:

```
【x-o-x.org】【44FB985C08_by_MEGAFUNPRO.COM】【17C8472BFA_by_OKFUN.ORG】【FCBZONEBDD4F0A112_by_FCBZONE.ORG】【HG4S83816NBVD_by_peru-boundary.com】【FASTJHM63C6BFA69NPG_by_FastZone.ORG】【倉庫 @http://txbpyc.com】
```

## Success
- BetterZip
- Entropy
- MacPAR Deluxe

## Fail
- Keka
- RAR Expander
- RAR Archive Utility
- StuffIt Expander

**Comment by gingerbeardman** [comment]
Looking at rar command line closer, I believe the problem is the <kbd>space</kbd> and possibly the following <kbd>@</kbd> characters in the password.

If you remove the <kbd>space</kbd> character, then the password is still valid and `rar` will extract it.

```
$ rar e test.rar -p【x-o-x.org】【44FB985C08_by_MEGAFUNPRO.COM】【17C8472BFA_by_OKFUN.ORG】【FCBZONEBDD4F0A112_by_FCBZONE.ORG】【HG4S83816NBVD_by_peru-boundary.com】【FASTJHM63C6BFA69NPG_by_FastZone.ORG】【倉庫@http://txbpyc.com

RAR 6.00 beta 1   Copyright (c) 1993-2020 Alexander Roshal   24 Oct 2020
Trial version             Type 'rar -?' for help


Extracting from test.rar

Extracting  test.txt                                                  OK 
All OK
```

## Success
- rar 5.91, 6.00b1 command line

## Fail
- Keka (repeatedly asks for password)
- RAR Expander ("The file 'test.rar' is not a RAR archive.")
- RAR-Archive Utility (fails silently; no error)
- StuffIt Expander ("The structure of the archive is damaged.")

**Comment by gingerbeardman** [comment]
Looking further, it seems that the whole long password may not required to decrypt the rar?

This password will extract the file using `rar` command line tool:

```
【x-o-x.org】【44FB985C08_by_MEGAFUNPRO.COM】【17C8472BFA_by_OKFUN.ORG 】【FCBZONEBDD4F0A112_by_FCBZONE.ORG】【HG4S83816NBVD_by_peru-boun
```

But will not work for the other apps that previously extracted the file with the full password.

I read the following:
> Please note that RAR maximum file encryption password length is 127 characters. Longer passwords are truncated to this length.

So I think there are many issues at play here.

**Comment by aonez** [comment]
Thanks for all the research and testing @gingerbeardman. You're right there, 127 truncated passwords. The other issue here is the symbol `【`

- Passing the password parameter it works: `rar e test.rar -p***`
- Using the pipe (not informing the `-p` argument) fails most probably breaks the coding: `rar e test.rar`

Keka uses the later, thus it fails. BetterZip (and probably the others) use XADMaster so no parameter involved there.

**Comment by aonez** [comment]
The symbols `【】` count two characters instead of one in the code that reads it (from stdin redirection). So the password is truncated way before it should. I've fixed this one and reported it to [RARLAB](https://www.rarlab.com).

Here a build to test: [Keka-v1.2.0-rc.1.r4158.7z](https://github.com/aonez/Keka/releases/download/dev-test-builds/Keka-v1.2.0-rc.1.r4158.7z)
Here the patch I've applied to fix this issue: [encoding-truncated-password.patch.zip](https://github.com/aonez/Keka/files/5474870/encoding-truncated-password.patch.zip)

**Comment by Gimmy1206** [comment]
Thanks! @aonez can fix this.
I don't need another rar archive app finally.

**Comment by liushulun** [comment]
> Hi @liushulun! I'm sorry I'm not familiar with the way Chinese characters are typed. What you need that Keka is lacking to write those characters? Maybe you can share a screencast showing what's needed?

Hi @aonez ! Sorry for waiting so long. I got this problem that, when inputing password while extracting or encrypting, keka doesn't allow me to change my input source to chinese pinyin, so that I can only type english characters, like the two screenshots below:

<img width="594" alt="encrypting" src="https://user-images.githubusercontent.com/38213534/97969491-18807280-1dfb-11eb-84b9-fe4d329fc385.png">

<img width="634" alt="extracting" src="https://user-images.githubusercontent.com/38213534/97969495-1b7b6300-1dfb-11eb-87ab-8eecf00b2201.png">

I can still copy the chinese characters and paste into the edit box, but not convenient though.

Hope chinese characters inputing could be supported later~

**Comment by gingerbeardman** [comment]
@liushulun do other password fields in macOS allow switching input source?

**Comment by aonez** [comment]
@gingerbeardman @liushulun just checked this is not possible in a standard password field. So I've created a new ticket to include an option to show the passwords (no using a secure field with asterisks) in #659.

Since this ticked evolved in to a RAR password fix I'm closing it and the original Chinese characters question will be fixed in #659 😉

**Comment by liushulun** [comment]
> @liushulun do other password fields in macOS allow switching input source?

That reminds me, thx!!!


> @gingerbeardman @liushulun just checked this is not possible in a standard password field. So I've created a new ticket to include an option to show the passwords (no using a secure field with asterisks) in #659.
> 
> Since this ticked evolved in to a RAR password fix I'm closing it and the original Chinese characters question will be fixed in #659 😉

It will helps a lot! As far as I know, this is the common way in chinese extract/encrypt software. Thx !!!

**Comment by aonez** [comment]
As a final followup, this is now also fixed in rarlab's `rar` and `unrar` binaries since version 6.00 beta 2.

Extracted from `whatsnew.txt`:

```txt
   Version 6.00 beta 2

   1. Bugs fixed:

      a) Unix RAR version failed to correctly process manually entered
         passwords if they did not exceed the maximum allowed 127 characters
         in UTF-32 encoding, but exceeded this threshold in UTF-8.
```
